src/lib/util/hash.cpp

```diff
@@ -27,48 +30,80 @@ char const *const hash_collection::HASH_TYPES_CRC_SHA1 = "RS";
 char const *const hash_collection::HASH_TYPES_ALL = "RS";
 
 
-
 //**************************************************************************
-//  HASH COLLECTION
+//  HASH CREATOR
 //**************************************************************************
 
-//-------------------------------------------------
-//  hash_collection - constructor
-//-------------------------------------------------
+namespace {
 
-hash_collection::hash_collection()
-	: m_has_crc32(false),
-		m_has_sha1(false),
-		m_creator(nullptr)
+class hash_creator : public write_stream
 {
-}
+public:
+	// constructor
+	hash_creator(hash_collection &hashes, bool doing_crc32, bool doing_sha1)
+		: m_hashes(hashes)
+	{
+		if (doing_crc32)
+			m_crc32_creator = std::make_optional<crc32_creator>();
+		if (doing_sha1)
+			m_sha1_creator = std::make_optional<sha1_creator>();
+	}
+
+	// add the given buffer to the hash
+	virtual std::error_condition write(void const *buffer, std::size_t length, std::size_t &actual) noexcept override
+	{
+		// append to each active hash
+		if (m_crc32_creator.has_value())
+			m_crc32_creator.value().append(buffer, length);
+		if (m_sha1_creator.has_value())
+			m_sha1_creator.value().append(buffer, length);
```

Comment:
Going out of one’s way to avoid using the `explicit operator bool` is an antipattern.  The idiomatic way to test whether something is dereferenceable in C++ is to test it for truthness – this works for naked pointers, smart pointers, `std::optional`, `std::function`, and anything else written in a sensible way.

Also, you shouldn’t be calling `.value()` after testing whether the optional contains a value, because it contains an additional check and a throw if the value isn’t set.  If you’ve already tested whether a value is present, you should use the dereference operators (`*` or `->`) to access the value and avoid the redundant check.  (You shouldn’t be using `.value()` in a `noexcept` function on principle, anyway.)